File: app/src/main/java/com/app/id/adapter/RecyclerViewAdapter.java
package com.app.id.adapter;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.app.id.R;
import com.app.id.view.AbstractView;

import java.util.ArrayList;
import java.util.List;

public abstract class RecyclerViewAdapter<M extends Parcelable, V extends AbstractView<M>> extends RecyclerView.Adapter<RecyclerViewAdapter.RecyclerViewHolder<V>> {

    private static final int LOADING_TYPE = 0;
    private static final int ITEM_TYPE = 1;

    LayoutInflater inflater;

    ArrayList<M> list = new ArrayList<>();

    boolean loading = false;

    public RecyclerViewAdapter(Context context) {
        inflater = LayoutInflater.from(context);
    }

    @Override
    public RecyclerViewHolder<V> onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == LOADING_TYPE) {
            return new RecyclerViewHolder<>(inflater.inflate(R.layout.list_item_loading, parent, false));
        } else {
            RecyclerViewHolder<V> holder = new RecyclerViewHolder<>(inflater.inflate(getLayoutResForViewType(viewType), parent, false));

            onPostCreateViewHolder(holder, parent);

            return holder;
        }
    }

    void onPostCreateViewHolder(RecyclerViewHolder<V> holder, ViewGroup parent) { }

    @LayoutRes
    abstract int getLayoutResForViewType(int viewType);

    @Override
    public void onBindViewHolder(RecyclerViewHolder<V> holder, int position) {
        if (holder.getItemViewType() == LOADING_TYPE) {
            RecyclerView.LayoutParams lp = (RecyclerView.LayoutParams) holder.itemView.getLayoutParams();

            if (list == null || list.size() == 0) {
                lp.height = lp.width = RecyclerView.LayoutParams.MATCH_PARENT;
            } else {
                lp.height = lp.width = RecyclerView.LayoutParams.WRAP_CONTENT;
            }
        } else {
            M m = list.get(position);

            bind(holder.get(), m);

            onPostBindViewHolder(holder, m, position);
        }
    }

    public M getLastItem() {
        if (list.isEmpty()) {
            return null;
        }

        return list.get(list.size() - 1);
    }

    public M getFirstItem() {
        if (list.isEmpty()) {
            return null;
        }

        return list.get(0);
    }

    void bind(V v, M m) {
        v.bind(m);
    }

    void onPostBindViewHolder(RecyclerViewHolder<V> holder, M m, int position) { }

    @Override
    public int getItemViewType(int position) {
        if (loading && position == (getItemCount() - 1)) {
            return LOADING_TYPE;
        }

        return getViewTypeForPosition(position);
    }

    int getViewTypeForPosition(int position) {
        return ITEM_TYPE;
    }

    public void update(M m) {
        int indexOf = list.indexOf(m);

        if (indexOf > -1) {
            notifyItemChanged(indexOf);
        }
    }

    public void setLoading(boolean loading) {
        boolean old = this.loading;

        this.loading = loading;

        if (old != loading) {
            int size = list == null ? 0 : list.size();

            if (loading)
                notifyItemInserted(size);
            else
                notifyItemRemoved(size);
        }
    }

    public boolean isLoading() {
        return loading;
    }

    public int getCount() {
        return list == null ? 0 : list.size();
    }

    public void clear() {
        int previousSize = list.size();

        list.clear();
        notifyItemRangeRemoved(0, previousSize);
    }

    public List<M> getList() {
        return list;
    }

    public void addAll(@NonNull List<M> list) {
        int previous = this.list.size();

        this.list.addAll(list);

        notifyItemRangeInserted(previous, list.size());
    }

    @Override
    public int getItemCount() {
        int count = getCount();

        if (loading) {
            count++;
        }

        return count;
    }

    public void onRestoreState(@NonNull Parcelable savedInstanceState) {
        if (savedInstanceState instanceof SavedState) {
            //noinspection unchecked
            SavedState<M> state = (SavedState<M>) savedInstanceState;

            list = state.list;
        }
    }

    public ArrayList<M> getItems() {
        return list;
    }

    public Parcelable onSaveInstanceState() {
        return new SavedState<>(list);
    }

    public static class SavedState<M extends Parcelable> implements Parcelable {
        public ArrayList<M> list;

        public SavedState(ArrayList<M> list) {
            this.list = list;
        }

        protected SavedState(Parcel in) {
            //noinspection unchecked
            list = in.readArrayList(RecyclerViewAdapter.class.getClassLoader());
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeList(list);
        }

        @Override
        public int describeContents() {
            return 0;
        }

        public static final Creator<SavedState> CREATOR = new Creator<SavedState>() {
            @Override
            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            @Override
            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

    public static class RecyclerViewHolder<V> extends RecyclerView.ViewHolder {
        public RecyclerViewHolder(View itemView) {
            super(itemView);
        }

        public V get() {
            //noinspection unchecked
            return (V) itemView;
        }
    }
}

File: app/src/main/java/com/app/id/adapter/RecyclerViewAdapter.kt
package com.app.id.adapter

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.app.id.R
import com.app.id.extensions.createParcel
import com.app.id.view.AbstractView
import java.util.*

abstract class RecyclerViewAdapter<M : Parcelable, V : AbstractView<M>>(context: Context) : RecyclerView.Adapter<RecyclerViewAdapter.RecyclerViewHolder<V>>() {

    internal var inflater: LayoutInflater

    var items: ArrayList<M> = ArrayList()
        internal set

    var isLoading = false
        set (loading) {
            val old = this.isLoading

            field = loading

            if (old != loading) {
                val size = items.size

                if (loading)
                    notifyItemInserted(size)
                else
                    notifyItemRemoved(size)
            }
        }

    init {
        inflater = LayoutInflater.from(context)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolder<V> {
        if (viewType == LOADING_TYPE) {
            return RecyclerViewHolder(inflater.inflate(R.layout.list_item_loading, parent, false))
        } else {
            return RecyclerViewHolder<V>(inflater.inflate(getLayoutResForViewType(viewType), parent, false)).apply {
                onPostCreateViewHolder(this, parent)
            }
        }
    }

    internal fun onPostCreateViewHolder(holder: RecyclerViewHolder<V>, parent: ViewGroup) { }

    @LayoutRes
    internal abstract fun getLayoutResForViewType(viewType: Int): Int

    override fun onBindViewHolder(holder: RecyclerViewHolder<V>, position: Int) {
        if (holder.itemViewType == LOADING_TYPE) {
            val lp = holder.itemView.layoutParams as RecyclerView.LayoutParams

            with (lp) {
                if (items.size == 0) {
                    height = RecyclerView.LayoutParams.MATCH_PARENT
                    width = RecyclerView.LayoutParams.MATCH_PARENT
                } else {
                    height = RecyclerView.LayoutParams.WRAP_CONTENT
                    width = RecyclerView.LayoutParams.WRAP_CONTENT
                }
            }
        } else {
            val m = items[position]

            bind(holder.get(), m)

            onPostBindViewHolder(holder, m, position)
        }
    }

    val lastItem: M?
        get() = items.lastOrNull()

    val firstItem: M?
        get() = items.firstOrNull()

    internal fun bind(v: V, m: M) {
        v.bind(m)
    }

    internal fun onPostBindViewHolder(holder: RecyclerViewHolder<V>, m: M, position: Int) { }

    override fun getItemViewType(position: Int): Int {
        if (isLoading && position == itemCount - 1) {
            return LOADING_TYPE
        }

        return getViewTypeForPosition(position)
    }

    internal fun getViewTypeForPosition(position: Int): Int = ITEM_TYPE

    fun update(m: M) {
        val indexOf = items.indexOf(m)

        if (indexOf > -1) {
            notifyItemChanged(indexOf)
        }
    }

    val count: Int
        get() = items.size

    fun clear() {
        val previousSize = items.size

        items.clear()
        notifyItemRangeRemoved(0, previousSize)
    }

    fun addAll(list: List<M>) {
        val previous = this.items.size

        this.items.addAll(list)

        notifyItemRangeInserted(previous, list.size)
    }

    override fun getItemCount(): Int {
        var count = count

        if (isLoading) {
            count++
        }

        return count
    }

    @Suppress("UNCHECKED_CAST")
    fun onRestoreState(savedInstanceState: Parcelable) {
        if (savedInstanceState is SavedState<*>) {
            items = savedInstanceState.let { it.list as? ArrayList<M>? } ?: ArrayList()
        }
    }

    fun onSaveInstanceState() = SavedState(items)

    data class SavedState<M: Parcelable> (var list: ArrayList<M>): Parcelable {
        override fun writeToParcel(dest: Parcel?, flags: Int) {
            dest?.writeTypedList(list)
        }

        override fun describeContents(): Int = 0

        protected constructor(parcelIn: Parcel): this(parcelIn.createTypedArrayList(SavedState.CREATOR) as? ArrayList<M> ?: ArrayList<M>())

        companion object {
            @JvmField @Suppress("unused")
            val CREATOR = createParcel { SavedState<Parcelable>(it) }
        }
    }

    class RecyclerViewHolder<V>(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @Suppress("UNCHECKED_CAST")
        fun get(): V = itemView as V
    }

    companion object {

        private val LOADING_TYPE = 0
        private val ITEM_TYPE = 1
    }
}
